### PR TITLE
Attempt to make compilation stack-safe

### PR DESCRIPTION
**Pull request by wemrysi**
Removes reliance on known sources of stack-unsafety in the compiler
  * scalaz.StateT
  * scalaz.Id
  * recursive scalaz.Equal for FreeMap

[ch11190]

### CODE REVIEW

**Comment by alissapajer** [comment]
🤞 really hope this does the trick!

**Comment by wemrysi** [comment]
@alissapajer it does the trick for the query in the referenced issue and removes all the problematic places I know of.

It is probably worth using `LazyEqual` for the other recursive types eventually, to be safe. `FreeMap`s are by far the largest structures we have since they are proportional to the number of columns, whereas the rest of `QScript` is proportional to the number of dimensional operations.

File: core/src/main/scala/quasar/compile/Compiler.scala [hunk 1]
```diff
@@ -39,14 +40,21 @@ import quasar.sql._
...
+
 import matryoshka._
 import matryoshka.data._
 import matryoshka.implicits._
 import matryoshka.patterns._
+
 import pathy.Path._
-import scalaz.{Plus => _, Tree => _, Free => ZFree, Select => _, _}, Scalaz._
+
+import scalaz.{\/, -\/, \/-, Cofree, Equal, Functor, Free => ZFree, Monad, Need, Scalaz, Show}, Scalaz._
```
Comment (nicflores) [inline core/src/main/scala/quasar/compile/Compiler.scala:52]: Dumb question over here. Why do we use Scalaz's monad instead of Cat's monad?

Comment (wemrysi) [reply core/src/main/scala/quasar/compile/Compiler.scala:52]: @nicflores In this case, just history. This is one of the oldest files in the codebase and we just haven't taken the time to port it over to cats. It also relies heavily on `matryoshka` which is based on scalaz.

**Pull request merged**